### ISSUE TITLE
Shortcut for group-level parameters in mixed models

### ISSUE THREAD

**Issue by DominiqueMakowski** [easystats/parameters#482]
Given the growth of interest and a posteriori re-usage of individual-level parameters, it would be nice to have a function that returns it directly, something like:

``` r
parameters_grouplevel <- function(model, ...) {
  out <- parameters::parameters(model, effects = "random", group_level = TRUE)
  out
}



model <- lme4::lmer(Reaction ~ Days + (1 + Days  | Subject), data = lme4::sleepstudy)

parameters_grouplevel(model)
#> # Random Effects
#> 
#> Parameter         | Coefficient |    SE |           95% CI
#> ----------------------------------------------------------
#> (Intercept) [308] |        2.26 | 12.07 | [-21.40,  25.92]
#> (Intercept) [309] |      -40.40 | 12.07 | [-64.06, -16.74]
#> (Intercept) [310] |      -38.96 | 12.07 | [-62.62, -15.30]
#> (Intercept) [330] |       23.69 | 12.07 | [  0.03,  47.35]
#> (Intercept) [331] |       22.26 | 12.07 | [ -1.40,  45.92]
#> (Intercept) [332] |        9.04 | 12.07 | [-14.62,  32.70]
#> (Intercept) [333] |       16.84 | 12.07 | [ -6.82,  40.50]
#> (Intercept) [334] |       -7.23 | 12.07 | [-30.89,  16.43]
#> (Intercept) [335] |       -0.33 | 12.07 | [-23.99,  23.32]
#> (Intercept) [337] |       34.89 | 12.07 | [ 11.23,  58.55]
#> (Intercept) [349] |      -25.21 | 12.07 | [-48.87,  -1.55]
#> (Intercept) [350] |      -13.07 | 12.07 | [-36.73,  10.59]
#> (Intercept) [351] |        4.58 | 12.07 | [-19.08,  28.24]
#> (Intercept) [352] |       20.86 | 12.07 | [ -2.79,  44.52]
#> (Intercept) [369] |        3.28 | 12.07 | [-20.38,  26.93]
#> (Intercept) [370] |      -25.61 | 12.07 | [-49.27,  -1.95]
#> (Intercept) [371] |        0.81 | 12.07 | [-22.85,  24.47]
#> (Intercept) [372] |       12.31 | 12.07 | [-11.34,  35.97]
#> Days [308]        |        9.20 |  2.30 | [  4.68,  13.72]
#> Days [309]        |       -8.62 |  2.30 | [-13.14,  -4.10]
#> Days [310]        |       -5.45 |  2.30 | [ -9.97,  -0.93]
#> Days [330]        |       -4.81 |  2.30 | [ -9.33,  -0.30]
#> Days [331]        |       -3.07 |  2.30 | [ -7.59,   1.45]
#> Days [332]        |       -0.27 |  2.30 | [ -4.79,   4.25]
#> Days [333]        |       -0.22 |  2.30 | [ -4.74,   4.29]
#> Days [334]        |        1.07 |  2.30 | [ -3.44,   5.59]
#> Days [335]        |      -10.75 |  2.30 | [-15.27,  -6.23]
#> Days [337]        |        8.63 |  2.30 | [  4.11,  13.15]
#> Days [349]        |        1.17 |  2.30 | [ -3.34,   5.69]
#> Days [350]        |        6.61 |  2.30 | [  2.10,  11.13]
#> Days [351]        |       -3.02 |  2.30 | [ -7.53,   1.50]
#> Days [352]        |        3.54 |  2.30 | [ -0.98,   8.05]
#> Days [369]        |        0.87 |  2.30 | [ -3.65,   5.39]
#> Days [370]        |        4.82 |  2.30 | [  0.31,   9.34]
#> Days [371]        |       -0.99 |  2.30 | [ -5.51,   3.53]
#> Days [372]        |        1.28 |  2.30 | [ -3.23,   5.80]
```

<sup>Created on 2021-05-04 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

Some possible additions could include:

- Adjust the printing method so that it doesn't merge the "Parameter" and the "Level" columns
- Adjust the printing method so that it doesn't drop the "Group" level
- Drop the `Effects` column (because it's always "random"?)
- Rearrange the df so that it starts with: `Group`, `Level`, and `Parameter`: since the focus is on the individual groups, the parameters should be organized by group and "by level" (?) - Even if we keep the arrangement by parameters (which also makes sense), the columns indicating the structure should be on the left-hand side
- Add a `reshape_*()` (or a `summary()`?) method that would rearrange the dataframe so that we have one row per group-level (and we add the parameter as a prefix): this would be super useful to get such a dataframe and then merge it with the initial dataframe

**Comment by bwiernik** [comment]
For comparison, this is the coef() function in the various mixed effects models

**Comment by mattansb** [comment]
Don't we have something like `get_random_effects()` in `insight`?

**Comment by strengejacke** [comment]
``` r
model <- lme4::lmer(Reaction ~ Days + (1 + Days  | Subject), data = lme4::sleepstudy)
insight::get_parameters(model, effects = "random")
#> $Subject
#>     (Intercept)        Days
#> 308   2.2585509   9.1989758
#> 309 -40.3987381  -8.6196806
#> 310 -38.9604090  -5.4488565
#> 330  23.6906196  -4.8143503
#> 331  22.2603126  -3.0699116
#> 332   9.0395679  -0.2721770
#> 333  16.8405086  -0.2236361
#> 334  -7.2326151   1.0745816
#> 335  -0.3336684 -10.7521652
#> 337  34.8904868   8.6282652
#> 349 -25.2102286   1.1734322
#> 350 -13.0700342   6.6142178
#> 351   4.5778642  -3.0152621
#> 352  20.8636782   3.5360011
#> 369   3.2754656   0.8722149
#> 370 -25.6129993   4.8224850
#> 371   0.8070461  -0.9881562
#> 372  12.3145921   1.2840221
#> 
#> with conditional variances for "Subject"
```

<sup>Created on 2021-05-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by strengejacke** [comment]
> For comparison, this is the coef() function in the various mixed effects models

I'm not sure it is `coef()` - it should be `ranef()`. coef is the fixef + ranef (the sum of coefficients).

**Comment by bwiernik** [comment]
Perhaps I misread, but I thought that sum was what Dominique was referring to

**Comment by DominiqueMakowski** [comment]
It seems my post was confusing ^^ Let me rephrase

Moving forward I can see an increase in the interest in group-level parameters (not only the coefs but also the SE CI etc). Currently afaik the canonical way to obtain them in easystats is `parameters::parameters(model, effects = "random", group_level = TRUE)`, but this is not as straightforward as it could (requiring arguments tweaking), and the presentation/printing is not optimal.

So I was thinking of adding essentially a shortcut to the above, that would be easily findable and useable. Also, we could improve the printing and add some more specific methods to do useful operations like smart reshaping and listifying.

It's just that as I'm currently using this functionality a lot (in the preparation of some teaching as well), I can see the benefit of having a specific function dedicated to query group-level parameters

**Comment by strengejacke** [comment]
> Adjust the printing method so that it doesn't merge the "Parameter" and the "Level" columns

Why not? That was actually intentional, to have a more compact overview.

> Adjust the printing method so that it doesn't drop the "Group" level

Not needed, as the output is split by groups.

> Drop the Effects column (because it's always "random"?)

That is intentional, to have compatibility to broom, so developers can easily replace broom by parameters, w/o having to adopt their code to a complete different behaviour.

> Rearrange the df so that it starts with: Group, Level, and Parameter: since the focus is on the individual groups, the parameters should be organized by group and "by level" (?) - Even if we keep the arrangement by parameters (which also makes sense), the columns indicating the structure should be on the left-hand side

Not sure if this will work, when you use `effects = "all"`, you get both fixed effects and random effects in _one_ data frame, so changing the order for random would affect the output for fixed as well.

> Add a reshape_*() (or a summary()?) method that would rearrange the dataframe so that we have one row per group-level (and we add the parameter as a prefix): this would be super useful to get such a dataframe and then merge it with the initial dataframe

This point could be addressed. :-)

You may look at this more complex example:

```r
library(parameters)
library(lme4)

data(sleepstudy)
set.seed(12345)
sleepstudy$grp <- sample(1:5, size = 180, replace = TRUE)
sleepstudy$subgrp <- NA
for (i in 1:5) {
  filter_group <- sleepstudy$grp == i
  sleepstudy$subgrp[filter_group] <-
    sample(1:30, size = sum(filter_group), replace = TRUE)
}
model <- lmer(
  Reaction ~ Days + (1 | grp / subgrp) + (1 | Subject),
  data = sleepstudy
)

parameters(model, effects = "all", group_level = TRUE)
```

**Comment by bwiernik** [comment]
> Given the growth of interest and a posteriori re-usage of individual-level parameters...

(Though personally I discourage folks from doing this, as it effectively treats the random parameter estimates as measured/known rather than estimated with uncertainty.)


> a `summary()` method

So it would likely look something like below. The issue is that just returning a data frame with the parameter labels, estimates, SEs, and grouping factors as columns makes it really difficult to merge back into the orignal data frame when there is more than one random effect factor because no row in the random effects data frame perfectly aligns with any row in the model frame. So, we would probably want a function that does the data merging for the user in one go.

``` r
library(parameters)
library(lme4)
#> Loading required package: Matrix
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tidyr)
#> 
#> Attaching package: 'tidyr'
#> The following objects are masked from 'package:Matrix':
#> 
#>     expand, pack, unpack
library(stringr)

data(sleepstudy)
set.seed(12345)
sleepstudy$grp <- sample(1:5, size = 180, replace = TRUE)
sleepstudy$subgrp <- NA
for (i in 1:5) {
    filter_group <- sleepstudy$grp == i
    sleepstudy$subgrp[filter_group] <-
        sample(1:30, size = sum(filter_group), replace = TRUE)
}
sleepstudy <- within(
    sleepstudy, 
    {
        Subject <- as.character(Subject)
        grp <- as.character(grp)
        subgrp <- as.character(subgrp)
    }
)
model <- lmer(
    Reaction ~ Days + (1 | grp / subgrp) + (1 | Subject),
    data = sleepstudy
)

reshaper <- function(x) {
    wider <- 
        pivot_wider(
            x,
            id_cols = c("Parameter", "Coefficient", "SE"),
            names_from = "Group",
            values_from = "Level"
        )
    if (any(str_detect(colnames(wider), ":"))) {
        column <- colnames(wider)[str_detect(colnames(wider), ":")]
        wider <- separate(
            wider, 
            col = column, 
            into = strsplit(column, ":")[[1]], 
            sep = ":", 
            remove = TRUE
        )
    }
    return(wider)
}
    



re_est <- parameters(model, effects = "random", group_level = TRUE) %>% 
    as.data.frame() %>% 
    select(Parameter, Group, Level, Coefficient, SE) %>%
    mutate(Parameter = paste(Parameter, Group, sep= "_")) %>% 
    split(., .$Group) %>% 
    lapply(reshaper)

augdata <- lapply(re_est, function(x) full_join(sleepstudy, x, by = colnames(x)[-c(1:3)])) %>%
    Reduce(rbind, .) %>% 
    pivot_wider(id_cols = -c(Parameter, Coefficient, SE),
                            names_from = "Parameter",
                            values_from = c("Coefficient", "SE")) %>% 
    as.data.frame()

head(augdata)
#>   Reaction Days Subject grp subgrp Coefficient_(Intercept)_grp
#> 1 249.5600    0     308   3     21                  -0.2506874
#> 2 258.7047    1     308   2      1                  -0.1641539
#> 3 250.8006    2     308   4     10                  -0.3914784
#> 4 321.4398    3     308   2     17                  -0.1641539
#> 5 356.8519    4     308   5     18                  -0.2001298
#> 6 414.6901    5     308   3      2                  -0.2506874
#>   Coefficient_(Intercept)_subgrp:grp Coefficient_(Intercept)_Subject
#> 1                        -1.73122252                        41.01657
#> 2                        -2.02950675                        41.01657
#> 3                        -2.80503086                        41.01657
#> 4                        -0.09515116                        41.01657
#> 5                         1.84368431                        41.01657
#> 6                         2.31306209                        41.01657
#>   SE_(Intercept)_grp SE_(Intercept)_subgrp:grp SE_(Intercept)_Subject
#> 1           1.747664                  6.110804               9.494543
#> 2           1.736175                  6.002382               9.494543
#> 3           1.747066                  5.900418               9.494543
#> 4           1.736175                  6.110737               9.494543
#> 5           1.753000                  5.921098               9.494543
#> 6           1.747664                  5.900352               9.494543
```

<sup>Created on 2021-05-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by bwiernik** [comment]
(And please forgive the use of `pivot_wider()` instead of `reshape()` in the above. `reshape()` makes me cry.)

**Comment by mattansb** [comment]
Reshape is horrible - @bwiernik , you are among friends 🥺

**Comment by bwiernik** [comment]
There is this recent thread on the r-devel listserv opining that maybe reshape isn't popular because of poor documentation: https://stat.ethz.ch/pipermail/r-devel/2021-April/080610.html

Don't really think that's the reason...

**Comment by DominiqueMakowski** [comment]
We should add a `data_to_wide()` and `data_to_long()` in insight that wraps around reshape but with the same flow as tidyr.

> Though personally I discourage folks from doing this, as it effectively treats the random parameter estimates as measured/known rather than estimated with uncertainty

Fair point, though to really address that I feel like there need to be some developments of statistics in which you can easily incorporate uncertainty. I mentioned this somewhere in easystats (and a consequence of this lost transmission can be seen here https://github.com/easystats/correlation/issues/28 ^^), but yeah I think that a tutorial / paper on "incorporate and propagate estimates uncertainty in statistics"  would be really helpful in general (far and beyond this issue here)

> Why not? That was actually intentional, to have a more compact overview. etc

regarding @strengejacke's reticence, let me emphasize that my proposal is not to modify any of the actual output / functions, but rather to provide a new function tailored and specific to group-level parameters (so no need for compatibility with broom nor consistency with fixed effects as this function would only return group-level parameters)

**Comment by strengejacke** [comment]
I still have no clear picture in mind of what you would like to have, so... would you mind, taking the lead for this issue?

**Comment by bwiernik** [comment]
There is this lean package that builds in a bunch of dplyr syntax with no dependencies. https://nathaneastwood.github.io/poorman/

It's been on my procrastination to do list to translate pivot_longer() and pivot_wider() using some of that package's tricks.

**Comment by bwiernik** [comment]
@strengejacke i think my reprex above is what @DominiqueMakowski is thinking

**Comment by DominiqueMakowski** [comment]
> There is this lean package that builds in a bunch of dplyr syntax with no dependencies. https://nathaneastwood.github.io/poorman/

Yes it's really nice. That said, we also could have our own data-wrangling functions with a few key design differences: 

1) don't overwrite tidyverse's name: I suppose many people use both verses so we don't want to user the same function names. Having them prefixed by `data_*` is IMO a good idea `data_select()`, `data_subset()`, etc.
2) Allowing both string input and NSE. Tidyverse's despite the recent iterations on the system is still hell to program with IMO, and besides that NSE is just one of this weird R thing (albeit very convenient), but very strange to explain to new students why sometimes stuff has to be quoted "because it's text not code" and in other cases "it's not quoted because well that's how it this you better remember it coz it doesn't work if you quote it". So having functions in which we can do `data_select(df, Sepal.Length)` and `data_select(df, "Sepal.Length")` would be awesome. That said it could be hassle to implement, so if we had to choose I'd definitely go for strings as a priority, it's just so much more easy to work with and manipulate (e.g., possibly of regex patterns etc.)
3) that's all I can think for now 😬 

> It's been on my procrastination to do list to translate pivot_longer() and pivot_wider() using some of that package's tricks.

We could add that to insight! it'd be useful immediately as we have `reshape_iterations()` in insight/bayestestR in which we could use it

**Comment by mattansb** [comment]
Hmmm someone should really start working on our data-package... 😁

**Comment by bwiernik** [comment]
1. One of poorman's design goals is interoperability with the dplyr API so that the packages can be swapped in scripts without any additional changes, so colliding with tidyverse names is a feature not a 🐛. Personally, I like the style of tidyverse and base's data manipulation functions being verbs, so I would prefer `reshape_longer()` and `reshape_wider()`.
2. I've implemented combo-NSE/SE in psychmeta, and poorman has written tricks to do most of the tidyselect helpers we could port. (As an aside: At this point as far as I am aware, all of the tidyverse functions will take names or strings for all of their arguments that accept names, so it never hurts to use string input, and when programming internally, they argue to never use NSE anyway and always using `.data` or `.env` to manage scope.)
3. For things other than reshape and perhaps summarize, base can already basically do all of the major data manipulation operations (e.g., see https://github.com/trinker/dplyr_in_a_nutshell). Functions like `with()`, `within()`, `subset()`, `transform()`, and `split()` aren't as widely known as they could (and should) be, but they exist and are pretty easy to use. And then there are also already tidyverse and data.table (and things like poorman for the dependency averse). 

So my thinking is rather that generally the R landscape for data manipulation tools is pretty robust and rich already. So, in contrast to working with statistical model results, which really need/needed attention from projects like easystats/broom, I don't really know how critical an easyverse data manipulation package is. Like you say, lots of folks use easystats+tidyverse or easystats+data.table.

**Comment by DominiqueMakowski** [comment]
let's move the discussion to #486
